components/application-role-mgt/org.wso2.carbon.identity.application.role.mgt/src/main/java/org/wso2/carbon/identity/application/role/mgt/ApplicationRoleManagerImpl.java

```diff
@@ -0,0 +1,244 @@
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.application.role.mgt;
+
+import org.apache.commons.lang.StringUtils;
+import org.wso2.carbon.identity.application.common.model.IdPGroup;
+import org.wso2.carbon.identity.application.common.model.IdentityProvider;
+import org.wso2.carbon.identity.application.role.mgt.dao.ApplicationRoleMgtDAO;
+import org.wso2.carbon.identity.application.role.mgt.dao.impl.ApplicationRoleMgtDAOImpl;
+import org.wso2.carbon.identity.application.role.mgt.dao.impl.CacheBackedApplicationRoleMgtDAOImpl;
+import org.wso2.carbon.identity.application.role.mgt.exceptions.ApplicationRoleManagementException;
+import org.wso2.carbon.identity.application.role.mgt.model.ApplicationRole;
+import org.wso2.carbon.identity.application.role.mgt.model.Group;
+import org.wso2.carbon.identity.application.role.mgt.util.ApplicationRoleMgtUtils;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.wso2.carbon.identity.application.role.mgt.constants.ApplicationRoleMgtConstants.ErrorMessages.ERROR_CODE_DUPLICATE_ROLE;
+import static org.wso2.carbon.identity.application.role.mgt.constants.ApplicationRoleMgtConstants.ErrorMessages.ERROR_CODE_INVALID_ROLE_NAME;
+import static org.wso2.carbon.identity.application.role.mgt.constants.ApplicationRoleMgtConstants.ErrorMessages.ERROR_CODE_ROLE_NOT_FOUND;
+import static org.wso2.carbon.identity.application.role.mgt.constants.ApplicationRoleMgtConstants.LOCAL_IDP;
+import static org.wso2.carbon.identity.application.role.mgt.util.ApplicationRoleMgtUtils.handleClientException;
+
+/**
+ * Application role management service implementation.
+ */
+public class ApplicationRoleManagerImpl implements ApplicationRoleManager {
+
+    private static final ApplicationRoleManager instance = new ApplicationRoleManagerImpl();
+
+    private ApplicationRoleManagerImpl() {
+
+    }
+
+    public static ApplicationRoleManager getInstance() {
+
+        return instance;
+    }
+
+    private final ApplicationRoleMgtDAO applicationRoleMgtDAO =
+            new CacheBackedApplicationRoleMgtDAOImpl(new ApplicationRoleMgtDAOImpl());
+
+    @Override
+    public ApplicationRole addApplicationRole(ApplicationRole applicationRole)
+            throws ApplicationRoleManagementException {
+
+        String tenantDomain = ApplicationRoleMgtUtils.getTenantDomain();
```

Comment:
Once role name validation is refactored into a private method, we won't need this variable.